### DISCUSSION TITLE
Replace qsort() with mutt_qsort_r()

### DISCUSSION THREAD

**Discussion by flatcap** [neomutt/neomutt#3952] [Development]
NeoMutt uses `qsort()` to sort lots of data.
It's quick, but some sorting requires extra data, which means global variables.

`qsort_r()` is an improved version that takes a data parameter, but it's not consistent across all platforms.

[`mutt_qsort_r()`](https://github.com/neomutt/neomutt/blob/a1913aa342687daa505d095152f4cb3fc8b5d92e/mutt/qsort_r.c#L65-L76) is a wrapper that hides the implementation details.

---

## Replace `qsort()` with `mutt_qsort_r()`

Below is a list of all the uses of `qsort()`, and their callback functions, including some used in `ARRAY_SORT()`.

Every callback function needs to be checked for:
- Global variables
  e.g. `static bool SbSortReverse = false;`
- Config variables
  Lookups are quick, but they should only be done once

[commands.c:818:5:](https://github.com/neomutt/neomutt/blob/main/commands.c#L818)                 `envlist_sort()`
[complete/helpers.c:390:5:](https://github.com/neomutt/neomutt/blob/main/complete/helpers.c#L390) `label_sort()`
[config/subset.c:94:3:](https://github.com/neomutt/neomutt/blob/main/config/subset.c#L94)         `elem_list_sort()`
[imap/imap.c:1570:3:](https://github.com/neomutt/neomutt/blob/main/imap/imap.c#L1570)             `imap_sort_email_uid()`
[maildir/maildir.c:234:7:](https://github.com/neomutt/neomutt/blob/main/maildir/maildir.c#L234)   `maildir_sort_flags()`
[ncrypt/dlg_gpgme.c:735:5:](https://github.com/neomutt/neomutt/blob/main/ncrypt/dlg_gpgme.c#L735) `crypt_compare_address_qsort()`, `crypt_compare_date_qsort()`, `crypt_compare_keyid_qsort()`, `crypt_compare_trust_qsort()`
[ncrypt/dlg_pgp.c:658:5:](https://github.com/neomutt/neomutt/blob/main/ncrypt/dlg_pgp.c#L658)     `pgp_compare_address_qsort()`, `pgp_compare_date_qsort()`, `pgp_compare_keyid_qsort()`, `pgp_compare_trust_qsort()`
[alias/sort.c:175:3:](https://github.com/neomutt/neomutt/blob/main/alias/sort.c#L175)             `alias_sort_address()`, `alias_sort_name()`, `alias_sort_name()`, `alias_sort_unsort()`
[browser/sort.c:214:3:](https://github.com/neomutt/neomutt/blob/main/browser/sort.c#L214)         `browser_compare()`
[core/command.c:59:3:](https://github.com/neomutt/neomutt/blob/main/core/command.c#L59)           `commands_sort()`
[imap/imap.c:1385:5:](https://github.com/neomutt/neomutt/blob/main/imap/imap.c#L1385)             `imap_sort_uid()`
[imap/imap.c:1485:5:](https://github.com/neomutt/neomutt/blob/main/imap/imap.c#L1485)             `imap_sort_uid()`
[imap/message.c:1653:3:](https://github.com/neomutt/neomutt/blob/main/imap/message.c#L1653)       `imap_sort_uid()`
[maildir/maildir.c:592:3:](https://github.com/neomutt/neomutt/blob/main/maildir/maildir.c#L592)   `maildir_sort_inode()`
[maildir/mh.c:674:5:](https://github.com/neomutt/neomutt/blob/main/maildir/mh.c#L674)             `mh_sort_path()`
[sidebar/sort.c:205:3:](https://github.com/neomutt/neomutt/blob/main/sidebar/sort.c#L205)         `sb_sort_count()`, `sb_sort_desc()`, `sb_sort_flagged()`, `sb_sort_order()`, `sb_sort_path()`, `sb_sort_unread()`

**Comment by roccoblues** [comment]
Do you want to replace all occurences or only the ones that actually would use the data parameter? For consistency it would probably be better to replace all?

**Reply by flatcap** [reply]
Yes, I think replacing all would be best

**Comment by roccoblues** [comment]
I can tackle this. Started here: https://github.com/neomutt/neomutt/pull/3953.

**Comment by flatcap** [comment]
Finished in: 02bd13556
